proposals/yield-in-try-catch.md

```diff
@@ -0,0 +1,388 @@
+
+### Code generation yield inside try with catch
+
+The code generation for iterators that have `yield` inside `try` blocks must preserve the same exception semantics for the `catch` in both `MoveNext` and `Dispose`. To achieve this the compiler will take a similar approach to what it does for `finally` blocks.
+
+For every `catch` block where the `try` has a nested `try / finally` with `yield`:
+
+1. The contents of the `catch` block will be generated into a parameterless method on the iterator with a `void` return.
+2. The contents of the `when` clause will be generated into a parameterless method on the iterator with a `bool` return.
+3. The `catch` block will be replaced with a call to the generated method in `MoveNext`.
+4. The `when` clause will be replaced with a call to the generated method in `MoveNext`.
+5. The `Dispose` method will mirror `catch` and `finally` blocks in the same way it mirrors `finally` blocks today and dispatch to the appropriate method
+
+The exception object will be lifted into a field just as any other local would be and accesses to it in the generated `catch` and `when` methods will use the field. In the case the language allowed `when` clauses the pattern variables would be lifted into fields as well.
+
+For example consider this code sample:
+
+```csharp
+static IEnumerable<int> Iterator()
+{
+    try
+    {
+        try
+        {
+            yield return M(1);
+        }
+        finally
+        {
+            Console.WriteLine("Finally Inner");
+        }
+    }
+    catch (InvalidOperationException ex1) when (ex.Message.Contains("hello"))
+    {
+        Console.WriteLine("Catch1");
+    }
+    catch (Exception ex2) 
+    {
+        Console.WriteLine("Catch2");
+    }
+    finally
+    {
+        Console.WriteLine("Finally Outer");
+    }
+}
+static T M<T>(T t) => t;
+```
+
+Will generate the following `Dispose` method:
+
+```csharp
+[DebuggerHidden]
+void IDisposable.Dispose()
+{
+    int num = <>1__state;
+    if ((uint)(num - -4) > 1u && num != 1)
+    {
+        return;
+    }
+    try
+    {
+        if (num != -4 && num != 1)
+        {
+            return;
+        }
+        try
+        {
+        }
+        finally
+        {
+            <>m__Finally2();
+        }
+    }
+    catch (InvalidOperationException ex) when (<>1__ex1 = ex, <>m__When1())
+    {
+        <>m__Catch1();
+    }
+    catch (Exception ex)
+    {
+        <>1__ex2 = ex;
+        <>m__Catch1();
+    }
+    finally
+    {
+        <>m__Finally1();
+    }
+}
+```
+
+The `<>1__ex1 = ex` in the `when` clause is not legal but the IL generated for the `when` will conceptually have this behavior.
+
+### Code generation yield inside try / catch in async iterators
+
+The code generation for `try / catch` blocks in async iterators will be largely the same as traditional iterators. The difference is that the return type of generated `catch` methods will be `ValueTask` instead of `void`.
```

Comment:
> The code generation for `try / catch` blocks in async iterators will be largely the same as traditional iterators.

I didn't follow this. Async iterators are more like async state machines than iterator state machines. 
Instead of generating a MoveNext and a Dispose method with similar structure, we use a single method (MoveNext) for both normal and disposal execution. A flag indicates if we're in disposal and we use it to skip over non-disposal code as appropriate.

Here are some notes on [disposal in async iterators.](https://github.com/jcouv/roslyn/blob/d0a56dda8423ad4ddd1f100c66ab56c0f9a343c3/async-iterator.md#design) Happy to discuss offline.

Reply:
Hmm. Maybe i was looking at the wrong code gen here. Will take a look at this.